### ISSUE TITLE
PIH dictionary import

### ISSUE THREAD

**Issue by paynejd** [OpenConceptLab/ocl_issues#45]

**Comment by lnball** [comment]
FYI @Karuhanga  

The PIHEMR dictionary has been cleaned up and in the OpenMRS dropbox.  When we are ready to try again, I can recreate the latest version.  There are some minor changes.

**Comment by paynejd** [comment]
FYI Rafal is working on the CIEL import now and we should be able to do it
via the new bulk import API, which means we will soon be able to run these
large imports on our own!

**Comment by jamlung-ri** [comment]
Has been validated against OpenMRS schema and can be imported. @bmamlin has been doing some importing on Staging. If this work is done, the ticket can be closed.

**Comment by jamlung-ri** [comment]
Moving to the next sprint since there are discrepancies remaining

**Comment by bmamlin** [comment]
The PIH dictionary was imported onto staging a couple months ago. This uncovered some issues with custom validation schema (#732), an issue with encoding of URLs for codes (#807), and the inability to delete a source if any mappings exist from other sources.

PIH-temp was imported a couple of weeks ago using the OpenMRS Custom Validation. After this import completed and an export of HEAD was downloaded, a few differences were discovered:

* One missing concept: "Dextrose 5%" (3779) on OCL.
  * This is one of five retired concepts. The other retired concepts (e.g., [802](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/802/)) were imported successfully.
  * I thought the percent sign (%) might be an issue, but there are 30 concepts with percent signs in their names.
* Two mappings that didn't match between import and OCL: [3779 &rarr; "Dextrose 5%"](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/mappings/2557764/) and [3779 &rarr; 7896](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/mappings/2554070/). Both are mappings from the concept that didn't import (3779). I think these didn't match between the import and OCL export because in these mappings on OCL the `from_concept_url` isn't populated (maybe because there was an error importing concept 3779?).
* Six mappings for the concept ["Time units" (466)](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/466/) are missing on OCL. I'm not yet sure why these failed to import.
  * 466 &ndash;<kbd>Q-AND-A</kbd>&rarr; [458](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/458/)
  * 466 &ndash;<kbd>Q-AND-A</kbd>&rarr; [464](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/464/)
  * 466 &ndash;<kbd>CONCEPT-SET</kbd>&rarr; [460](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/460/)
  * 466 &ndash;<kbd>CONCEPT-SET</kbd>&rarr; [461](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/461/)
  * 466 &ndash;<kbd>CONCEPT-SET &rarr;</kbd>[462](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/462/)
  * 466 &ndash;<kbd>CONCEPT-SET</kbd>&rarr; [463](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/463/)

<details><summary>Click for full diff</summary>

```json
{
  "missing_concepts":[
    {"retired": true, "datatype": "N/A", "type": "Concept", "concept_class": "Misc", "source": "PIH-temp", "extras": {}, "names": [{"locale": "fr", "external_id": "3e754086-55c0-4862-8a50-26e0b6f8f512", "locale_preferred": true, "name": "Dextrose 5% fluide", "name_type": "FULLY_SPECIFIED"}, {"locale": "en", "external_id": "5c702054-6a5c-11e2-b6f9-aa00f871a3e1", "locale_preferred": true, "name": "Dextrose 5%", "name_type": "FULLY_SPECIFIED"}], "owner": "PIH", "owner_type": "Organization", "external_id": "5c6b1834-6a5c-11e2-b6f9-aa00f871a3e1", "id": "3779", "descriptions": []}
  ],
  "extra_concepts": [],
  "missing_mappings": [
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/466/", "external_id": "0b24a340-15f5-102d-96e4-000c29c2a5d7", "source": "PIH-temp", "owner": "PIH", "map_type": "Q-AND-A", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/458/"},
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/466/", "external_id": "5013CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC", "source": "PIH-temp", "owner": "PIH", "map_type": "Q-AND-A", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/464/"},
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/466/", "external_id": "1865AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE", "source": "PIH-temp", "owner": "PIH", "map_type": "CONCEPT-SET", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/461/"},
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/466/", "external_id": "1862AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE", "source": "PIH-temp", "owner": "PIH", "map_type": "CONCEPT-SET", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/460/"},
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/466/", "external_id": "1860AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE", "source": "PIH-temp", "owner": "PIH", "map_type": "CONCEPT-SET", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/463/"},
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/466/", "external_id": "1864AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE", "source": "PIH-temp", "owner": "PIH", "map_type": "CONCEPT-SET", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/462/"},
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/3779/", "external_id": "5c703940-6a5c-11e2-b6f9-aa00f871a3e1", "source": "PIH-temp", "owner": "PIH", "map_type": "SAME-AS", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/Dextrose+5%25/"},
    {"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/3779/", "external_id": "5c706ce4-6a5c-11e2-b6f9-aa00f871a3e1", "source": "PIH-temp", "owner": "PIH", "map_type": "SAME-AS", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH-temp/concepts/7896/"}
  ],
  "extra_mappings": [
    {"external_id": "5c703940-6a5c-11e2-b6f9-aa00f871a3e1", "retired": false, "map_type": "SAME-AS", "source": "PIH-temp", "owner": "PIH", "owner_type": "Organization", "from_concept_code": "3779", "from_concept_name": null, "from_concept_url": "", "to_concept_code": "Dextrose 5%", "to_concept_name": null, "to_concept_url": null, "from_source_owner": "PIH", "from_source_owner_type": "Organization", "from_source_url": "/orgs/PIH/sources/PIH-temp/", "from_source_name": "PIH-temp", "to_source_owner": "PIH", "to_source_owner_type": "Organization", "to_source_url": "/orgs/PIH/sources/PIH-temp/", "to_source_name": "PIH-temp", "url": "/orgs/PIH/sources/PIH-temp/mappings/2557764/", "version": "2557769", "id": "2557764", "versioned_object_id": 2557764, "versioned_object_url": "/orgs/PIH/sources/PIH-temp/mappings/2557764/", "is_latest_version": true, "update_comment": null, "version_url": "/orgs/PIH/sources/PIH-temp/mappings/2557764/2557769/", "uuid": "2557769", "version_created_on": "2021-07-25T15:32:26.538712Z", "from_source_version": null, "to_source_version": null, "extras": {}, "type": "Mapping", "created_on": "2021-07-25T15:32:26.538712Z", "updated_on": "2021-07-25T15:32:26.548985Z", "created_by": "ocladmin", "updated_by": "ocladmin", "internal_reference_id": "2557769"},
    {"external_id": "5c706ce4-6a5c-11e2-b6f9-aa00f871a3e1", "retired": false, "map_type": "SAME-AS", "source": "PIH-temp", "owner": "PIH", "owner_type": "Organization", "from_concept_code": "3779", "from_concept_name": null, "from_concept_url": "", "to_concept_code": "7896", "to_concept_name": null, "to_concept_url": null, "from_source_owner": "PIH", "from_source_owner_type": "Organization", "from_source_url": "/orgs/PIH/sources/PIH-temp/", "from_source_name": "PIH-temp", "to_source_owner": "PIH", "to_source_owner_type": "Organization", "to_source_url": "/orgs/PIH/sources/PIH-temp/", "to_source_name": "PIH-temp", "url": "/orgs/PIH/sources/PIH-temp/mappings/2554070/", "version": "2554073", "id": "2554070", "versioned_object_id": 2554070, "versioned_object_url": "/orgs/PIH/sources/PIH-temp/mappings/2554070/", "is_latest_version": true, "update_comment": null, "version_url": "/orgs/PIH/sources/PIH-temp/mappings/2554070/2554073/", "uuid": "2554073", "version_created_on": "2021-07-25T15:30:53.725502Z", "from_source_version": null, "to_source_version": null, "extras": {}, "type": "Mapping", "created_on": "2021-07-25T15:30:53.725502Z", "updated_on": "2021-07-25T15:30:53.739779Z", "created_by": "ocladmin", "updated_by": "ocladmin", "internal_reference_id": "2554073"}
  ]
}
```
</details>

**Comment by snyaggarwal** [comment]
@bmamlin The concept 3779 failed to import because of locales -- https://github.com/OpenConceptLab/ocl_issues/issues/732#issuecomment-886247072

I will check the mappings.

**Comment by bmamlin** [comment]
Thanks @snyaggarwal. That makes sense. I'd like to confirm (with Andy & Jon) that this name uniqueness constraint should apply across retired concepts (e.g., if I want to retire a concept and replace it with an improved concept with the same name, I'd be forced to rename the old retired concept before I could add the new one).

**Comment by lnball** [comment]
@bmamlin I think the issue is that the concept "Dextrose 5% (PIH:7896)" is retired but the names are not voided.  I guess this is correct behavior but OCL doesn't like it.  It conflicts with the 'en' synonym of "Dextrose 5% (CIEL:161250)".  There could be data for the retired concept in various deployments so I cannot delete the retired one.

I can write liquibase to migrate the data, update any forms that used the old concept, then delete the old concept.  Let me know if that would be helpful.

**Comment by snyaggarwal** [comment]
@bmamlin ok that make sense. I will update the validator.

**Comment by bmamlin** [comment]
@snyaggarwal did we figure out the issue with the 6 mappings that didn't go in? Is OCL preventing multiple mappings between the same two concepts?

@lnball my hunch is "Time Units" (PIH 6412) is an oddball concept that is both coded (with answers) _and_ has concept set members (even though it's not a concept set based on its datatype).

**Comment by lnball** [comment]
I have cleaned up PIH:6412 to remove the sets.  CIEL gets away with many shinanigans (see screenshot):

![image](https://user-images.githubusercontent.com/1164849/128217615-db68771a-97a7-42e5-9171-1eff6bcaf959.png)

Let me know if you figure out the rest.  I can generate a new file for import whenever you are ready.  @bmamlin @snyaggarwal

**Comment by snyaggarwal** [comment]
@bmamlin Yes for OpenMRS schema, there can be only [one mapping between two concepts](https://github.com/OpenConceptLab/oclapi2/blob/3c846a54b7979fb6fbf3bf99f154f46273550730/core/mappings/custom_validators.py#L15).

**Comment by bmamlin** [comment]
@lnball go ahead and generate a new file and we'll give it a whirl. 🙂

**Comment by bmamlin** [comment]
@lnball I began adapting the `sql_to_json` script to use the gold mapping (a SAME-AS mapping to a numeric PIH code) and discovered several cases where there is more than one gold mapping for a concept. For example, from the latest version you sent there are 16 cases where concepts have two PIH "gold" IDs:

|id|gold mappings|
|:-:|:-:|
|6|5622, 6408
|7|9721, 1915
|461|1074, 2161
|462|2160, 1073
|758|1900, 1907
|1121|2475, 1118
|2041|7789, 161283
|2807|2763, 7644
|2935|1459, 1897
|2951|1500, 1191
|2952|1190, 1499
|3159|2074, 2097
|3271|3519, 987
|3366|1546, 2181
|3665|4042, 4043
|4789|2858, 2853

I was thinking we could use these "gold mappings" to define the ID of the concept within OCL; however, that doesn't work if there are multiple gold mappings. I understand why there are some PIH SAME-AS mappings to non-numeric codes (e.g., so modules, forms, and reports can use friendly references like "PIH:ASTHMA"); however, I'm unclear why there would be multiple numeric codes assigned for the same concept. Does this represent an artifact from concepts that were retired/changed over time? Would it be possible to constrain gold mappings (i.e., a SAME-AS mapping to a PIH numeric code) to one per concept?

**Comment by bmamlin** [comment]
For the record, @lnball discovered some of the duplicate "gold mappings" exist in CIEL, so sent a list to @askanter for fixing in CIEL as well:
![image](https://user-images.githubusercontent.com/860834/129287133-56025849-3de4-44e5-914d-be76c440d942.png)

**Comment by lnball** [comment]
@bmamlin If it wasn't clear on the emails with Andy, he added PIH mappings to CIEL concepts many years ago.  Many of those PIH mappings are out-of-date, wrong, etc.  When I use mds to bring concepts from CIEL, those mapping come for free :(

Because it's quick, I'll generate a new mysqldump of the PIH concepts today.  

But I think you'd like the numeric PIH mappings on all PIH concepts so you can use them for concept ids?  If that's true, I will add them even where there's a CIEL mapping.  This will take a bit longer to write the script but especially since it would require regenerating all 24 mds packages...  

If it's helpful to do this right, I'm game.

**Comment by askanter** [comment]
Just to add my comment from the email here. CIEL maintains multiple maps between concepts when the source concepts change. For example, AMPATH might have changed ids over time for WEIGHT. There could be three different concept IDs in the AMPATH dictionary (two being retired) that have instance data attached to them. If we were migrating that dictionary to CIEL, we'd want to know that all of that instance data should be assigned to the CIEL concept for WEIGHT. 

So, I was hoping that we could keep all the maps, but perhaps identify one of them as the preferred map for the IDs you are creating. That would be the one that is currently active in the PIH dictionary. The mappings on the CIEL concepts should eventually be updated to indicate preference for outgoing maps (from CIEL to preferred PIH or AMPATH map)

**Comment by lnball** [comment]
@askanter What do you think about using a mapping relationship for the old ones?  "Associated with"?  Or a new relationship/name?

**Comment by askanter** [comment]
I will do so review. I know SNOMED has to do something for retired maps. But knowing which are retired in CIEL will be an update process for sure.

**Comment by mogoodrich** [comment]
@bmamlin was is the specific requirements for a "gold" mapping?  Would it be better to generate some sort of new mapping at import time instead of jumping through hoops making sure the PIH mappings are unique and on all concepts (And fair number of our concepts don't have numeric PIH mappings as all)?  I'm conscious of @lnball spending hours needing to manually add mappings and clean up duplicates.

What is the underlying structure/requirements in OCL that we need to fulfill?

**Comment by bmamlin** [comment]
> @bmamlin what is the specific requirements for a "gold" mapping? Would it be better to generate some sort of new mapping at import time instead of jumping through hoops making sure the PIH mappings are unique and on all concepts (And fair number of our concepts don't have numeric PIH mappings as all)? I'm conscious of @lnball spending hours needing to manually add mappings and clean up duplicates.
> 
> What is the underlying structure/requirements in OCL that we need to fulfill?

The problem is the SQL dumps from Ellen have non-official concept IDs, so when I imported the PIH concepts, they end up in OCL with those random concept IDs – e.g., "PIH:3" is "officially" the Asthma concept, but the concept in the SQL dump with `concept_id` 3 is a concept for a complex obs attachment and the Asthma concept is something like 730 with a SAME-AS mapping to PIH:3.

So, I thought we could use these SAME-AS mapping to PIH numeric codes (i.e., anything matching `/^[0-9]+$/`, meaning any whole number) as "gold mappings" to identify the "official" concept ID within the PIH organization.

Unfortunately, that has failed for a few reasons:
* Not all PIH concepts have one of these "gold mappings" and the random IDs can conflict with them (there were >600 conflicts like this in the last PIH SQL dump)
* Some concepts (16 of them) have two "gold mappings", making the "official" concept ID ambiguous
* CIEL contains some of these "gold mappings" to PIH, including cases where there are multiple for a concept that @askanter would prefer not to delete

While I think the approach of creating numeric (and non-numeric) SAME-AS mappings within an organization to be able to find the right concepts across servers (where internal concept IDs might vary) is a best practice that we should encourage across all orgs, the whole notion of using these as "gold mappings" (using them to identify the "official" concept ID) isn't going to work. That means that Ellen would need to give me a SQL dump where the concept IDs were the actual "official" concept IDs for PIH – i.e., a SQL dump from your "gold" concept server.

Fundamentally, I can't import the PIH dictionary without a non-ambiguously knowing the concept ID of each concept. The simplest solution (from my end) would be to use a SQL dump where concept IDs are "correct" (e.g., concept 3 is the Asthma concept). If that's hard for Ellen/you to generate, then I'm happy to meet in the middle with a scripted solution.

**Comment by mogoodrich** [comment]
Maybe we should talk at the next OCL call (can you remind me when it is :).

Since we started using Metadata Sharing years ago, PIH concepts no longer have "official" concept_ids, beyond the uuids, so it's not possible to "non-ambiguously know the concept ID of each concept", and we shouldn't rely on anything in the "concept_id" column to be "official"..  When we originally created the PIH numeric mappings, we used the concept_id as an easy way to generate a mapping, but there is no longer any expectation that this would stay in sync with the concept_id column in the OpenMRS DB.

If the use case is that OCL needs a primary key, can't we just generate that as part of the import script?

If the idea is that it's best practice that that primary key in OCL is also a mapping on the concept, perhaps we could create a new concept reference source like "PIH_OCL_CODE" and create a script to assign a code from that source to all concepts in the PIH EMR, and then you could use that code as the primary key upon input?

Sorry if I'm missing something... happy to talk more next week.





If

**Comment by bmamlin** [comment]
Well, @lnball provided a snapshot with gold mappings (a SAME-AS mapping to a numeric PIH code) for every concept. Initially, I thought I could simply swap in the gold mapping code as the concept IDs as I created the import file. But, as it turns out, every reference to the internal ID needs to be changed to the gold mapping (including Q&A or concept set references from other concepts and within the gold mapping's `from_concept_url`).

So, I tweaked the `ocl_omrs` script to gather all gold mappings and then substitute them, when appropriate, for all concept IDs and any mapping references to the org's concept IDs. I uploaded an initial pass of this onto staging as `PIH-temp` and, while the counts aren't an exact match (there are still a few issues to work out), it should be good enough for testing out with the subscription module.

**Comment by bmamlin** [comment]
Currently blocked by #1018.

**Comment by bmamlin** [comment]
I'm going to work on scripting the creation of the PIH-temp collection for now.

**Comment by mogoodrich** [comment]
Cool, definitely ping me @bmamlin  when that is ready...

**Comment by mogoodrich** [comment]
Just posted this on Slack:  In OpenMRS, when you associate answer concepts with a question, you can control the order via a "sortWeight" parameter (which can be used, for instance, to control the order of answers in a dropdown).  It looks like in OCL question-and-answer relationships are handled using a "Q-AND-A" mapping, and I don't see any sort weight or ordering ability on mappings?

This might be a blocker for wider use.

**Comment by mseaton** [comment]
@bmamlin let me know if it would make any sense for @mogoodrich or I to get up to speed on the process you have been using to load (or reload) the PIH dictionary into OCL staging, and whether or not either of us could take this if we end up needing to iterate on this.  It seems like the next step is to tweak the scripts you are using to ensure that sort weight is populated in the extras property for Q-AND-A and CONCEPT-SET concept mappings based on @mogoodrich comments.

**Comment by bmamlin** [comment]
> @bmamlin looks like we will need to modify the importer to set the sort_weight on the "extras" parameter for mappings of type "Q-AND-A" and "CONCEPT-SET"

I made this change [here](https://github.com/OpenConceptLab/ocl_omrs/commit/a6947b92fd38d8dc7571fc4ba1bc6cbaa1c9d637) and regenerated the import JSON for PIH-temp, which includes sort weights. I tried doing a bulk import of one mapping as a test and it didn't add the sort weight as an extra as expected, so I'm working through that now.

**Comment by bmamlin** [comment]
It turns out there's a bug for bulk import that doesn't update mappings as expected unless the import is done in parallel mode. @snyaggarwal pointed this out and I was able to import all answer & set member mappings to `PIH-temp` on staging, so OCL has `sort_weight` for all PIH answers & set members for `PIH-temp` now.

**Comment by mogoodrich** [comment]
just to confirm @bmamlin , this means that from your perspective, the sort_weights have now been added and we are ready to test again?  (though from the OCL module side we will need to make sure https://issues.openmrs.org/browse/OCLOMRS-1067 is implemented first before testing)

**Comment by bmamlin** [comment]
> from your perspective, the sort_weights have now been added and we are ready to test again?

Yes. For example, here's the ["Ambulatory" answer for "Type of patient"](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/mappings/4327041/):

<img src="https://user-images.githubusercontent.com/860834/138380254-e983b188-e459-4f73-a2a4-665eaae4cdf3.png" width=50% height=50%>

**Comment by mseaton** [comment]
@bmamlin - just following up to check if you saw @mogoodrich 's question and issue above?

**Comment by mogoodrich** [comment]
Thanks @mseaton ... I pinged Burke again on Slack today as well and he responded:

"@mogoodrich OCLOMRS-1067 looks like an OCLM ticket. It looks like sort_weight support for mappings already exists and is used for answers, but not yet for set members. I added a comment on OCLOMRS-1067 to this effect. Would you be able to tweak the Saver add set member method to use the same approach as the add answer method (pull sort weight from mapping)?"

(Not sure I follow, but will dig deeper once I'm up on email, just wanted to paste this here for reference before I forget).

**Comment by mseaton** [comment]
@bmamlin / @mogoodrich - on Slack?  I don't see anything.  Was it a PM?  I don't understand the above.  I agree that OCLOMRS-1067 is an OCLM ticket, and should be moved there.  But that isn't the issue we are describing above.  What you are describing above is that - when you download a zip file containing the json of a Collection from OCL, and inspect it in a text editor, the "extras" is empty where we expect to see the sort order.  Wouldn't that be an issue with the OCL exporter?

**Comment by mogoodrich** [comment]
Thanks @mseaton ... you are correct, that's why I was originally confused by @bmamlin 's response above and wanted to reread again in case I missed something.  My ping on Slack was likely confusing and I did it on the wrong post...

@bmamlin  any thoughts on the above?

**Comment by bmamlin** [comment]
Ah. Sorry about the confusion. I saw the comment in Slack referring to [OCLOMRS-1067](https://issues.openmrs.org/browse/OCLOMRS-1067) and so was responding to Mark's comment as if he was blocked by that issue.

You're correct. The `sort_weight` isn't in the export for the collection. It looks like this is another example of why we need the dynamic references that the OCL team is actively adding to the platform. Here's what happened:
* I created a collection from the PIH source. Currently (without dynamic references), references within a collection are static, pointing to a specific version of a resource.
* This made a static reference within the collection to the mapping _before_ the sort weight was added. If you look within the [actual reference in the collection](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH-temp/references?q=4327041&isTable=true&isList=false&isSplit=false&page=1&exactMatch=off), you can see it points to a specific version of the mapping: [/orgs/PIH/sources/PIH-temp/mappings/4327041/4327043/](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/mappings/4327041/4327043/) (the `4327043` at the end is a version number)

It can be confusing, since the Term Browser shows concepts & mappings in a collection using links to head (not the version in its references).

The near-term solution is to recreate the collection (updating its references to the latest versions). I'll do this now and then report back here.

Ultimately, this will be improved because the collection – with the upcoming dynamic references in OCL – will be able to have a reference like "all concepts & mappings from the source" along with an "auto expand" feature that will automatically use the latest version.

**Comment by bmamlin** [comment]
I deleted the PIH-temp collection on staging, recreated it, and have started a task to add references to all concepts & mappings within the PIH-temp source:

```
DELETE /orgs/PIH/collections/PIH-temp/

PUT /orgs/PIH/collections/" @pih-temp-collection.json
{
    "id": "PIH-temp",
    "short_code": "PIH-temp",
    "name": "PIH-temp",
    "full_name": "PIH-temp",
    "preferred_source": "CIEL",
    "public_access": "None",
    "collection_type": "Dictionary",
    "custom_validation_schema": "OpenMRS",
    "supported_locales":"en,am,bn,es,fr,ht,id,it,nl,pt,ru,rw,sw,ti,ur,vi"
}

PUT /orgs/PIH/collections/PIH-temp/references/
{
  "data": {
    "uri": "/orgs/PIH/sources/PIH-temp/",
    "concepts": "*",
    "mappings": "*"
  }
}
```

The task is started and will hopefully finish running overnight:

```
GET /manage/bulkimport/?task=b368c8e7-3bad-4145-ac7f-ec12fc1872b0
{
    "queue": "default",
    "state": "STARTED",
    "task": "b368c8e7-3bad-4145-ac7f-ec12fc1872b0",
    "username": ""
}
```

**Comment by mogoodrich** [comment]
Thanks Burke!

**Comment by mogoodrich** [comment]
Also, @bmamlin , until dynamic references are added, that means that each time we want to update a concept, we need to recreate the entire PIH collection?  This would seem like a must-have for MVP for PIH (though it sounds like the OCL team is currently working on it).

**Comment by mseaton** [comment]
Thanks @bmamlin - I just downloaded the PIH-temp Collection version 2.0, and I can see the sortWeight property under extras, so we'll report back how we get on with this.

**Comment by ibacher** [comment]
> we need to recreate the entire PIH collection?

Not really. It's just a matter of updating the reference to that concept in the collection; the rest of the collection can remain as it is. Dynamic references should make that easier though.

**Comment by mseaton** [comment]
@bmamlin I commented on the Talk post that there a handful of Concepts in varying states that didn't make it into the final Collection export.  These may not all be the same issue.  Can we investigate what happened here?

Some these seem to have made it correctly into the PIH-temp source, but are not in the PIH-temp Collection for some reason:
* [Urinary Cast, Hyaline](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/12299/)
* [Pregnancy](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/44/)
* [Torsion of Ovary](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/10133/)
* [Triage green](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/8887/)
* [Sodium](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/12658/)
* [Negative](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/664/)
* [Yellow](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/10667/)
* [Triage red](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/8886/)
* [Protein](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/12659/)
* [Case status](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/7085/)
* [Urinalysis](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/302/)
* [Glucose Tolerance Test](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/11967/)

Others appear to be in CIEL, but I don't see them in the PIH-temp source:
* [HIV resulting in infectious or parasitic disease](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/160159/)
* [Patient’s family history list](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/160593/)
* [Abacavir sulfate](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/70057/)
* [Very low density lipoprotein (mmol/L)](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/1298/)
* [Low-density Lipoprotein Cholesterol (mmol/L)](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/1008/)

**Comment by mseaton** [comment]
@bmamlin the mappings that ended up in the PIH-temp dictionary have many issues.  The Q-AND-A and CONCEPT-SET mappings seem fine (though obviously missing concepts lead to missing answers and set members).

The PIH mappings seem the most problematic.  There are many examples of concepts getting associated with completely the wrong concept.  I have to assume this is the result of issues with our multitude of PIH / PIH-temp / etc sources and such, but defer to you.  Some examples:

* [Fluoxetine](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/4045/):  has a SAME-AS mapping to PIH-temp:10916.  
* [Glibenclamide ](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/4046/):  has a SAME-AS mapping to PIH-temp:10820
* etc. etc.  many examples of this.

There are also many examples of missing mappings that I would have expected.  For example:

* [Minutes](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/6901/) - missing SNOMED CT:258701004

Finally, I'm curious about self-mappings.  I don't see mappings back to self for a lot of the PIH concepts.  I could see not adding these, but this seems to result in no PIH mappings getting created on these concepts when importing via openconceptlab module.  And it seems like other concepts do have these self-mappings.  [For example this one](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/1008/).

These are a few of the many examples of concepts missing self mappings:

* [Acute rheumatic fever](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/3308/) - MISSING MAPPING TO PIH:3308
* [Pit latrine ](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH-temp/concepts/11649/)- missing mapping to PIH:11649

**Comment by bmamlin** [comment]
> I don't see mappings back to self for a lot of the PIH concepts.

Up until now, I believe the import script was skipping over self mappings. As we're adding support for gold mappings, we should start to see these in OCL.

### Status update
* I've received an updated copy of the PIH dictionary (20211201) with gold mappings
* I added a [use_gold_mappings option](https://github.com/OpenConceptLab/ocl_omrs/commit/c0186da906caea146964db12afd80e04e6dbb051) to the `ocl_omrs` import script
* I've successfully generate the import json for the updated PIH dictionary
* Sunny made progress on #661 (ability to delete sources) and, while it's not yet possible to delete a large source through the UI (until the UI has async support #1128), it can now be done asynchronously through the API, so `PIH-temp` is no more and the existing (old) PIH dictionary on staging is in the process of being deleted.
* Once the PIH source has been successfully deleted, I'll import the latest version (20211201) so PIH can do further testing on it.

**Comment by mogoodrich** [comment]
Sweet, thanks @bmamlin !

**Comment by bmamlin** [comment]
Looking at discrepancies between the PIH 2021-12-01 source and what ended up in OCL, there was one concept that didn't get imported: "Baptist" (10149). The reason was its UUID (external_id) `1162932AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA` is 37 chars and our OpenMRS custom validation schema enforces a 36-character maximum. While I thought it would be from one too many A's, it looks like there's actually an extra leading "1" compared to CIEL's [Baptist Religion](https://app.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/162932/) concept (UUID `162932AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA`). I'm going to manually fix this in my import file for now, but PIH will want to correct it in their dictionary.

There were 5,786 mappings that didn't get created for some reason.
* It looks like 1,105 of these are gold mappings. OCL is refusing to create these mapping, reporting the error "Parent, map_type, from_concept, to_source, to_concept_code must be unique." For example, concept 1065 is mapped to itself with `{"retired": false, "from_concept_url": "/orgs/PIH/sources/PIH/concepts/1065/", "external_id": "bbb5c16f-f127-32a4-97d5-7d7bed12bfd4", "source": "PIH", "owner": "PIH", "map_type": "SAME-AS", "owner_type": "Organization", "type": "Mapping", "to_concept_url": "/orgs/PIH/sources/PIH/concepts/1065/"}` and this fails to get imported. I'll ask the OCL devs about it.
* There are 4,681 mappings that appeared to be missing in the export from OCL, yet OCL reported as "updated" when I tried to re-upload them. I've queued a new export to see if they're in it or still missing. If they exist in the new export, maybe it's possible I managed to generate the initial export before these mappings were fully indexed.

Let me see if I can get (at least most of) these issues sorted out before PIH tries testing sync.

**Comment by lnball** [comment]
Good sleuthing @bmamlin .  I will change the uuid for Baptist to drop the extra '1' prefix.

Keep me posted on the other missing maps.

**Comment by bmamlin** [comment]
Okay. I think I discovered the problem (or at least a significant chunk). I adapted the import script to use gold mappings for _mappings_; however, I forgot to do the same for concept sets and answers. Each of these becomes a mapping in OCL and the concept ID referenced on either side (concept set or member, question or answer) needs to be switched from the concept ID in OpenMRS to the proper gold mapping code within the import json. I'll need to adapt the import script, re-generate the import file, and import PIH again to fix this.

**Comment by mseaton** [comment]
Awesome, thanks @bmamlin !

**Comment by bmamlin** [comment]
I've regenerated the import file and triggered deletion of the PIH source on staging. Once it's deleted, I'll re-import a fixed version of the PIH dictionary and re-validate to see if we are (hopefully) more on the goodrich... er, I mean mark.

**Comment by bmamlin** [comment]
PIH has been deleted from staging. I will plan on trying the re-import tomorrow afternoon/evening.

**Comment by bmamlin** [comment]
PIH source on staging has been updated. _Much_ closer this time. Concept count matches exactly. Mappings are off by 13, but it looks like 13 were "updated" by OCL, so maybe there are some duplicates. Investigating...

FYI - I haven't created a matching collection yet, so not quite yet ready for PIH to do testing.

**Comment by bmamlin** [comment]
Almost there. The 13 mappings that failed to import are all mappings for the MDR-TB module. I think OCL is incorrectly considering these equivalent to gold mappings. I reported this on #1151. If this can be easily fixed, then I'll make one final pass to import the PIH dictionary and expect it will be successful.

In the meantime, I went ahead and created a PIH collection, so PIH can proceed with testing (understanding the thirteen MDR-TB mappings will be missing for now).

<details><summary><em>Click to see <code>org.openmrs.module.mdrtb</code> mappings that did not import...</em></summary>

| Concept | Code           |
|--------:|----------------|
|    703  | POSITIVE       |
|    656  | ISONIAZID      |
|     58  | TUBERCULOSIS   |
|    740  | CIPROFLOXACIN  |
|   6349  | RELAPSE        |
|   1413  | CYCLOSERINE    |
|    755  | LEVOFLOXACIN   |
|   2459  | CLARITHROMYCIN |
|   745   | ETHAMBUTOL     |
|   2460  | RIFABUTINE     |
|   5829  | PYRAZINAMIDE   |
|    766  | PYRIDOXINE     |
|   1412  | CLOFAZIMINE    |
</details>

**Comment by lnball** [comment]
@bmamlin we are no longer using the mdrtb module or those mappings.  I'm happy to remove from our golden dictionary.  Did you want me to regenerate a new mysqldump without these 13 mappings?

FYI @mogoodrich

**Comment by lnball** [comment]
@bmamlin Too late, Burke.  When I didn't hear back, I deleted all those mappings ;) and regenerated the PIH EMR dictionary

**Comment by mseaton** [comment]
@bmamlin - I had a chance to do some testing with this today and things are looking much, much better.  There is still an issue with around 12 concepts not making it into the final collection somehow.  These are _mostly_ but not _entirely_ the same 12 concepts as I reported earlier in this thread.  When I import the OCL package zip into a clean DB, I get errors trying to import mappings that refer to these concepts, and the concepts don't exist.  And all of these concepts should exist.  They are as follows:
```
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/10678/ - Mobility
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/790/ - Creatinine (umol/L)
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/1006/ - Total cholesterol
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12763/ - ABC
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12299/ - Urinary Cast, Hyaline
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/44/ - Pregnancy
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/8887/ - Green
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/664/ - Negative
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12659/ - Protein
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/7085/ - Case status
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/302/ - Urinalysis
https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/11967/ - Glucose tolerance test
```

As you can see, all of these concepts were successfully imported into the PIH _source_.  But for whatever reason they are not part of the PIH _collection_.  So I'm not really sure this is an issue with the import tool, or a broader OCL bug, or something else - thoughts?

Note, I also find it strange that in OCL, it says that the PIH Collection contains 5942 concepts, yet the database that we provided as a source contains 5960 concepts, which is then only 5948 concepts due to the 12 that are not imported that are listed above.  Any idea why OCL lists the collection as 5942 and not 5948?  See version 1.0.0 that I created here:  https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/versions

**Comment by snyaggarwal** [comment]
@mseaton OCL 5942 are active concepts and 6 are retired concepts. If you include retired ones the count is correct [5948](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=&isTable=true&isList=false&isSplit=false&page=1&exactMatch=off&includeRetired=true)

**Comment by bmamlin** [comment]
> As you can see, all of these concepts were successfully imported into the PIH _source_. But for whatever reason they are not part of the PIH _collection_. So I'm not really sure this is an issue with the import tool, or a broader OCL bug, or something else - thoughts?

Interesting. I created the PIH collection using:

```
PUT /orgs/PIH/collections/PIH/references/
{
  "data": {
    "expressions": [
      "/org/PIH/sources/PIH/concepts/*",
      "/org/PIH/sources/PIH/mappings/*"
    ]
  }
}
```

which _should_ include all concepts and mappings from the PIH source within the collection. As a first step, I'll try deleting the PIH collection and recreating it to see if the problem is reproducible.

**Comment by bmamlin** [comment]
Well, it looks like re-creating the PIH collection brought us back to the same place... so it's recreatable. Interestingly, it looks like, while all 12 are missing from the collection, 5 of the 12 concepts exist when fetched directly via the API. I'm not seeing an obvious pattern.

| Source | Collection in App | API |
|-|-|-|
| [Mobility](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/10678/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=10678&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/10678/) |
| [Creatinine (umol/L)](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/790/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=790&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/790/) |
| [Total cholesterol](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/1006/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=1006&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Missing](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/1006/) |
| [ABC](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12763/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=12763&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/12763/) |
| [Urinary Cast, Hyaline](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12299/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=12299&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Missing](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/12299/) |
| [Pregnancy](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/44/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=44&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/44/) |
| [Green](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/8887/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=8887&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Missing](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/8887/) |
| [Negative](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/664/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=664&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/664/) |
| [Protein](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12659/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=12659&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Missing](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/12659/) |
| [Case status](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/7085/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=7085&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Missing](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/7085/) |
| [Urinalysis](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/302/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=302&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Missing](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/302/) |
| [Glucose tolerance test](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/11967/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=11967&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Missing](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/11967/) |

@snyaggarwal, any idea why the 12 concepts above aren't being seen in the PIH collection on staging? I created them using [the API call mentioned above](https://github.com/OpenConceptLab/ocl_issues/issues/45#issuecomment-998233934) to add all concepts & mappings to the collection (twice now) and these 12 are consistently misbehaving. I can find five of them as concepts through the collection using the API, the other seven are missing, and all 12 are missing from the collection in the app.

**Comment by lnball** [comment]
Not obvious to me.  You have questions, drug, misc, test.  Assorted mappings.  Thanks for the sleuthing @bmamlin

**Comment by snyaggarwal** [comment]
@bmamlin Below are the reported failures for PIH collection add_references task:
```
{
"task-id": "677ca08f-3238-47ec-95df-2f89587129b9",
"state": "SUCCESS",
"result": {
"/orgs/PIH/sources/PIH/concepts/8887/5561799/": [
"Concept preferred name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/302/5563616/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/12299/5563483/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/7085/5564293/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/12060/5572237/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/7896/5572611/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/12270/5563020/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/1006/5563457/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/3188/5572841/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/11967/5563402/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/12659/5573081/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/10133/5569173/": [
"Concept fully specified name must be unique for same collection and locale."
],
"/orgs/PIH/sources/PIH/concepts/11689/5563408/": [
"Concept preferred name must be unique for same collection and locale."
]
}
}
```

**Comment by bmamlin** [comment]
@snyaggarwal, thanks. That's interesting, since the PIH source also has the OpenMRS custom validation schema. If these were duplicates, I wouldn't expect them to have been allowed into the source. Do collections and sources use the same custom validation logic? If so, how could entries in a collection pass validation in the source and fail in the collection?

**Comment by snyaggarwal** [comment]
@bmamlin Ya the OpenMRS validation for collection is not exactly the same -- [here is the code](https://github.com/OpenConceptLab/oclapi2/blob/7b27a0c1ebaa00526ba36320bed057e66f42fcbf/core/collections/models.py#L133)

**Comment by bmamlin** [comment]
Thanks @snyaggarwal. I think we're getting closer to the issue. I took one of the examples from your list of errors: the concept "Green". The PIH source contains [Green (8887)](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/8887/) and [Green color (12572)](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12572/). The latter has a short name "Green". This is allowed, since the duplicate constraint for fully specified names and preferred names should only be checking within fully specified names and preferred names of the same locale, respectively. In other words, if one concept has a preferred name of "Green" in the `en` locale, then no other concept can also have a preferred name of "Green" in the `en` locale; however, it's okay for another concept to have a short name or a non-preferred synonym of "Green" in the `en` locale.

I believe the fix for this with source validation might be in [this line](https://github.com/OpenConceptLab/oclapi2/blob/7b27a0c1ebaa00526ba36320bed057e66f42fcbf/core/concepts/custom_validators.py#L78), scoping the search to the specific name type. Can we do the same for collections? I believe this will solve our problem.

**Comment by snyaggarwal** [comment]
Yep working on it.

**Comment by snyaggarwal** [comment]
@bmamlin This is deployed on Staging now.

**Comment by mogoodrich** [comment]
Awesome, thanks so much @bmamlin  @snyaggarwal  @lnball

**Comment by bmamlin** [comment]
Thanks @snyaggarwal. I've deleted the PIH collection on staging and am recreating it now. 🤞

**Comment by bmamlin** [comment]
@snyaggarwal we seem to be much closer. The concept & mapping counts are now exact matches between PIH collection & source on Staging 🥳; however, some of the concepts don't show up in the UI as expected. For each of the following seven concepts, they seem to be missing from the collection (they don't show up as expected in the UI), yet they can be retrieved via the API:

| Source | Collection in App | API |
|-|-|-|
| [Total cholesterol](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/1006/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=1006&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/1006/) |
| [Urinary Cast, Hyaline](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12299/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=12299&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/12299/) |
| [Green](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/8887/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=8887&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/8887/) |
| [Protein](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/12659/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=12659&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/12659/) |
| [Case status](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/7085/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=7085&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/7085/) |
| [Urinalysis](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/302/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=302&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/302/) |
| [Glucose tolerance test](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/11967/) | [Missing](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=11967&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) | [Exists](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/11967/) |

For example, the API returns [`/orgs/PIH/collections/PIH/concepts/8887/`](https://api.staging.openconceptlab.org/orgs/PIH/collections/PIH/concepts/8887/) with `"id": "8887"` but a search within the PIH collection for an exact match to 8887 [comes up empty](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=8887&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on). **Any idea what might explain that behavior?**

BTW, I tried to get a full report of the task from `/importers/bulk-import/?task=a53bc8dd-e9b7-4225-884f-d77fc82bbc27&result=json`, but only get this back instead of a full list of successes & failures as expected:
```json
{
    "queue": "default",
    "state": "SUCCESS",
    "task": "a53bc8dd-e9b7-4225-884f-d77fc82bbc27",
    "username": ""
}
```

**Comment by snyaggarwal** [comment]
@bmamlin its may be because the indexing is still happening.

**Comment by snyaggarwal** [comment]
@bmamlin I can see [8887](https://app.staging.openconceptlab.org/#/orgs/PIH/collections/PIH/concepts/?q=8887&isTable=true&isList=false&isSplit=false&page=1&exactMatch=on) now. I would wait for indexing to finish. I will update here when that gets done.

**Comment by bmamlin** [comment]
Ah. Awesome! It looks like it was an indexing delay after all. Everything is showing up as expected now!

**Comment by bmamlin** [comment]
@mseaton @mogoodrich take another look at PIH on Staging. The dictionary (source & collection) are looking pretty good. Throw some rocks at it and see if it looks good. If it does, we can consider the import process as "working" and we can start tackling other issues (source handling, sync, etc.).

**Comment by lnball** [comment]
What is this?  The not-link symbol says "Target concept is not defined in OCL".

![image](https://user-images.githubusercontent.com/1164849/147115808-d3f2e90f-6654-405d-8614-206d5a0ddc4c.png)

**Comment by bmamlin** [comment]
> What is this? The not-link symbol says "Target concept is not defined in OCL".

OCL term browser automatically creates a link for mappings to concepts that exist in OCL. Mappings to things that don't exist in OCL (like the code "Triage green") can't be linked. So, term browser is just saying "I can't take you to this concept." Showing the inability to link makes a little more sense when the mapping is to a code in a standard (like LOINC or ICD) or a concept in another implementation's dictionary that doesn't happen to be loaded into OCL rather than a mapping like this to an free text code PIH is using for convenience.

**Comment by mseaton** [comment]
> @mseaton @mogoodrich take another look at PIH on Staging

Success!!  Things are looking good @bmamlin - thanks!  The number of concepts match exactly, and my comparisons only find a few issues, all of which are known and generally considered low priority or a non-priority.  Those are that we lose concept version, concept retire reason, and voided concept names in the process of moving our concepts to OCL.  

So yes, I think we can consider this "working" and that we have successfully imported  the PIH dictionary into OCL and can get it back out again and imported into a new OpenMRS instance with the OCL subscription module.

Tackling the issue of source handling sounds like a good next area to focus on.  Thanks again for all of the help and work on this.

**Comment by mseaton** [comment]
Just noticed we are coming up on the 4 year anniversery for this ticket - great to see it getting so close to resolution!

**Comment by bmamlin** [comment]
> Success!! Things are looking good @bmamlin - thanks!

Thanks so much, @mseaton. Given this success, I'm going to close this ticket to acknowledge the fundamental work of being able to import PIH into OCL has been successfully completed. We can track remaining steps of getting to a place where PIH can use OCL to manage their dictionaries through ongoing [discussions in Talk](https://talk.openmrs.org/t/migrating-pih-dictionary-to-ocl/34981/) and other tickets. Done in less than 4 years! 😄

**Comment by paynejd** [comment]
Wahooo! Congratulations everyone- it's been a long time coming ;)

**Comment by lnball** [comment]
4 years!  A new record???

Thanks so much, team.
